geaflow/geaflow-dsl/geaflow-dsl-connector/geaflow-dsl-connector-paimon/src/main/java/org/apache/geaflow/dsl/connector/paimon/PaimonTableSource.java

```diff
@@ -89,71 +90,65 @@ public void init(Configuration tableConf, TableSchema tableSchema) {
             String optionJson = tableConf.getString(PaimonConfigKeys.GEAFLOW_DSL_PAIMON_OPTIONS_JSON);
             Map<String, String> userOptions = GsonUtil.parse(optionJson);
             if (userOptions != null) {
-                for (Map.Entry<String, String> entry : userOptions.entrySet()) {
-                    options.put(entry.getKey(), entry.getValue());
-                }
+                options.putAll(userOptions);
             }
             this.configJson =
                 tableConf.getString(PaimonConfigKeys.GEAFLOW_DSL_PAIMON_CONFIGURATION_JSON, "");
             if (!StringUtils.isBlank(configJson)) {
                 Map<String, String> userConfig = GsonUtil.parse(configJson);
                 if (userConfig != null) {
-                    for (Map.Entry<String, String> entry : userConfig.entrySet()) {
-                        configs.put(entry.getKey(), entry.getValue());
-                    }
+                    configs.putAll(userConfig);
                 }
             }
         }
         this.database = tableConf.getString(PaimonConfigKeys.GEAFLOW_DSL_PAIMON_DATABASE_NAME);
         this.table = tableConf.getString(PaimonConfigKeys.GEAFLOW_DSL_PAIMON_TABLE_NAME);
+        this.sourceMode = SourceMode.valueOf(tableConf.getString(PaimonConfigKeys.GEAFLOW_DSL_PAIMON_SOURCE_MODE).toUpperCase());
     }
 
     @Override
     public List<Partition> listPartitions(int parallelism) {
-        return listPartitions();
+        List<Partition> partitions = new ArrayList<>(parallelism);
+        for (int i = 0; i < parallelism; i++) {
+            partitions.add(new PaimonPartition(database, table));
+        }
+        return partitions;
+    }
+
+    @Override
+    public List<Partition> listPartitions() {
+        return listPartitions(1);
```

Comment:
Implementing listPartitions() with a default parallelism of 1 is a suboptimal design. Throwing an exception would be a more explicit and safe approach.

Reply:
This implementation is quit similar with FileTableSource. I've deep dive into the invocation, GeaFlowTableSourceFunction and GeaFlowQueryEngine will call listPartitions(parallelism) rather listPartitions().